Add checkpoints to timestepping methods

Psirus opened the pull request.

Psirus committed:
Add checkpoints to timestepping methods

TTitscher commented on fenics_helpers/timestepping.py:
```diff
@@ -55,25 +55,37 @@ def __init__(self, solve, post_process, u):
         progress = Progress(t_start, t_end, show_bar)
         while t < t_end:
-            t += dt
+            if checkpoints is not None and checkpoints[current_checkpoint] < t + dt:
```
What is the expected behavior for `t_start = 0` and `checkpoints=[-1]`? 
Currently, this would mean starting at `t = -1`, which is bad IMO.

I would expect that the time stepping includes the checkpoints. So silently excluding them, is also bad.

What do think about adding 
~~~py
assert(checkpoints[0] >= t_start)
assert(checkpoints[-1] <= t_end)
~~~
after `np.sort`?

TTitscher commented on fenics_helpers/timestepping.py:
```diff
@@ -105,9 +117,12 @@ def __init__(self, solve, post_process):
-        for t in np.append(np.arange(t_start, t_end, dt), t_end):
+        points_in_time = np.hstack((np.arange(t_start, t_end, dt), t_end))
+        if checkpoints is not None:
+            points_in_time = np.hstack((points_in_time, checkpoints))
```
Same as above: IMO checkpoints should be between t_start and t_end.

TTitscher commented on fenics_helpers/timestepping.py:
```diff
@@ -105,9 +117,12 @@ def __init__(self, solve, post_process):
+    def run(self, t_end, dt, t_start=0.0, checkpoints=None, show_bar=False):
         progress = Progress(t_start, t_end, show_bar)
-        for t in np.append(np.arange(t_start, t_end, dt), t_end):
+        points_in_time = np.hstack((np.arange(t_start, t_end, dt), t_end))
```
Why `hstack` over `append`?

TTitscher commented on tests/test_timestepping.py:
```diff
@@ -1,13 +1,49 @@
+        pp = lambda t: visited_timesteps.append(t)
+        self.adaptive._post_process = pp
+        self.assertTrue(self.adaptive.run(1.5, dt0=1.0, checkpoints=[1.05, 1.25]))
+        self.assertTrue(np.count_nonzero(np.isin(visited_timesteps, [1.05, 1.25])) == 2)
```
`self.assertEqual(..., 2)` prints the values in case of a failure. I always use that when comparing numbers.

Psirus replied:
Sure. Copypasta mistake.

TTitscher reviewed (approved):
Nice! This is also very useful for my fatigue simulations, where I want to be sure that the max and min of a cycle is properly resolved. 

Idk about the issue with checkpoints outside of `t_start` and `t_end`. I would prefer a check, but whatever.

Psirus committed:
Add hypothesis testing to timestepping

Psirus committed:
Add hypothesis to CI configs

Psirus committed:
Use append instead of hstack

Psirus committed:
Move checkpoints into a separate class

TTitscher commented on fenics_helpers/timestepping.py:
```diff
@@ -42,7 +42,32 @@ def __del__(self):
+    def check(self, t, dt):
+        if self.points.size != 0 and self.points[0] < t + dt:
+            dt = self.points[0] - t
+            t = self.points[0]
```
I was wondering for a long time, why we modify t here. Then I realized the `else` branch also modifies it. It's a small detail, but I would prefer the method to only return a modified `dt`. Maybe also rename to `timestep`. The call side would then look like
~~~
dt = checkpoints.timestep(t, dt)
t += dt
~~~

Then the actual advancement of the global time is still in the `TimeStepping`, not in the checkpoint. That way, restoring the previous time with `t -= dt` would be more clear.

TTitscher commented on fenics_helpers/timestepping.py:
```diff
@@ -42,7 +42,32 @@ def __del__(self):
+    def __init__(self, points, t_start, t_end, dt_min=1e-6):
+        self.points = np.array(points)
+        self.post_process_t0 = False
+        if self.points.size != 0:
```
We could return early with
~~~
if not self.points:
    return
~~~
to flatten the method.

Psirus committed:
Revert changing current time in checkpoints

Psirus commented:
... also removed the FEniCS 2017.2 tests. Didn't seem worth it.

TTitscher commented:
Good point. Setting dt_min = 1.e-7 in the time integration will cause the tests to fail... I'll think about it

TTitscher committed:
Refactor output in class Progress

TTitscher committed:
Use dt in adaptive stepping right away (no dt0)

TTitscher committed:
Renaming. Remove 'TimeStepping' from classes.

And remove equidistant/adaptive in test method names

TTitscher committed:
Add type checks for solve method

Psirus commented on fenics_helpers/timestepping.py:
```diff
@@ -6,42 +6,39 @@ class Progress:
+            self._pbar = tqdm(total=t_end - t_start, ascii=True, bar_format=fmt)
+
+    def _green(self, msg):
+        info("\033[32m" + msg + "\033[m")
```
Not sure I like these methods. For one, they aren't really methods (as in `self` is not used), so they probably should be free functions. For two, they do two things at once, make stuff green and print.

I'd say we either just use string concatenation (no need to have this in this class though)
~~~python
RED = "\033[31m"
GREEN = "\033[32m"
RESET = "\033[m"
print(RED + msg + RESET)
~~~
or write a small function so that you can do
~~~python
print(colored(msg, "red"))
~~~

Psirus commented:
Like the renaming and extra tests & asserts! :+1:

TTitscher committed:
Remove dt_min from the checkpoint class

All checkpoints are now visited, regardless of dtmin. Additionally, time
steps that are set because of checkpoints do not influcence the normal
time step. Imagine
it0: dt = 1
it1: dt = 1.e-5 (to reach the checkpoint)
it2: dt = 1 (return to the previous one)

TTitscher committed:
Separate info/tqdm, add colored(msg, color_name)

TTitscher committed:
Add tqdm to CI. remove pointless t, dt0 = ..., dt0

to t = ...

TTitscher committed:
Fix further floating point precision issues

TTitscher committed:
Fix overshooting at t = t_end

Add TimeStepping class as wrapper to unify the API for easier change of
the time stepping scheme.

TTitscher commented:
I added a classs `TimeStepping` with methods `adaptive` and `equidistant` that wraps Adaptive.run or Equidistant.run. Do you think that this belongs here?

TTitscher committed:
Fix bug: Checkpoint steps did not decrease dt

Causes infinite loop as dt is tried over and over again. Add
DeterministicSolve to the test case to test it properly.

Psirus committed:
Change non-empty check to work with numpy arrays

Also made TERM_COLOR a dictionary.

TTitscher committed:
Unite Adaptive/Equidistant to TimeStepper

TTitscher commented on fenics_helpers/timestepping.py:
```diff
@@ -3,15 +3,11 @@
-        return "\033[31m"
-    if color_name == "green":
-        return "\033[32m"
+TERM_COLOR = {"red": "\033[31", "green": "\033[32"}
```
So much better. Automatic key errors, etc...
I never think of dicts...

TTitscher committed:
Add dt_max/min as method that support chaining

stepper = TimeStepper(...)
stepper.dt_min = 0.1
stepper.adaptive(1, ...)

becomes

TimeStepper(...).dt_min(0.1).adaptive(1,...)
